### DISCUSSION TITLE
Translatable plugin locale switcher not working

### DISCUSSION THREAD

**Discussion by christopheraseidl** [filamentphp/filament#11542] [Help]
### Package

Other

### Package Version

v3.2.35

### How can we help you?

I have installed the Spatie Translatable plugin v3.2, followed the instructions in the Filament and Spatie documentation, and set up my resource to translate columns on a relationship called "meta." I have two problems at the moment:

1. The locale switcher defaults to English, whereas my current app locale is actually Spanish. How do I get the locale switcher to sync with what my current locale is?

2. When I switch locales, the translatable fields all go blank. Looking in the database, the data is there, exactly like the English data (added via a Seeder). E.g. for my title column: `{"en": "Course", "es": "Curso"}`. Here is the relevant area in the form:

```
Section::make('Details')
    ->relationship('meta')
    ->schema([
        TextInput::make('title')
    ]);
```

**Comment by leandrocfe** [comment]
Did you add the meta in the $translatable property in the model?

**Reply by christopheraseidl** [reply]
I did, as per Spatie's documentation. I'm going to try to isolate where this problem is coming from. I'm sure my implementation is wrong somewhere.

**Reply by christopheraseidl** [reply]
OK, so I've detailed my setup below, which doesn't work, probably because I'm not setting the relationship up right for the plugin. Here are the issues I have:

1. I can create a Post, but when the form is submitted, only the translations for the form's `$activeLocale` are entered into the database.
2. This point is the cause of point 1; switching between locales, the form doesn't "remember" the values for the other fields. This problem is solved _only on the Create form_ if I change the Post model's `$translatable` array to just read 'meta,' but then I get a MySQL error because the translatable plugin is trying to do this, interpreting what I understand as the "meta" relationship as another column on the "posts" table:

```
insert into `posts` (`meta`, `updated_at`, `created_at`) values ({"en":{"title":"Title","slug":"title","description":"<p>description<\/p>","image":[],"meta_type":"App\\Models\\Post"}}, 2024-02-24 12:00:09, 2024-02-24 12:00:09)
```
3. The edit page doesn't switch between locales correctly no matter what I do, and tries to perform this operation on form submission, which causes an error:

```
update `posts` set `meta`.`title` = {"en":""}, `meta`.`slug` = {"en":""}, `meta`.`description` = {"en":""}, `posts`.`updated_at` = 2024-02-24 12:04:36 where `id` = 2
```

Post:

```
class Post extends Model
{
    use \Spatie\Translatable\HasTranslations;

    protected $with = [
        'meta',
    ];

    public $translatable = [
        'meta.title',
        'meta.slug',
        'meta.description',
    ];

    public function meta(): MorphOne
    {
        return $this->morphOne(Meta::class, 'meta');
    }
}
```

Meta:

```
class Meta extends Model
{
    use \Spatie\Translatable\HasTranslations;

    protected $fillable = [
        'title',
        'slug',
        'description',
        'image',
        'meta_id',
        'meta_type',
    ];

    protected $table = 'meta';

    public $translatable = [
        'title',
        'slug',
        'description',
    ];

    public function meta(): morphTo
    {
        return $this->morphTo();
    }

}
```

PostResource:

```
class PostResource extends Resource
{
    use \Filament\Resources\Concerns\Translatable;

    protected static ?string $model = Post::class;

    protected static ?string $navigationIcon = 'heroicon-o-rectangle-stack';

    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Section::make('Details')
                    ->relationship('meta')
                    ->schema([
                        TextInput::make('title'),
                        TextInput::make('slug'),
                        Forms\Components\RichEditor::make('description'),
                        FileUpload::make('image')
                            ->image(),
                    ]),
            ]);
    }

    public static function getPages(): array
    {
        return [
            'index' => Pages\ListPosts::route('/'),
            'create' => Pages\CreatePost::route('/create'),
            'edit' => Pages\EditPost::route('/{record}/edit'),
        ];
    }

    public static function getRelations(): array
    {
        return [
            //
        ];
    }

    public static function table(Table $table): Table
    {
        return $table
            ->columns([
                //
            ])
            ->filters([
                //
            ])
            ->actions([
                Tables\Actions\EditAction::make(),
                Tables\Actions\DeleteAction::make(),
            ])
            ->bulkActions([
                Tables\Actions\BulkActionGroup::make([
                    Tables\Actions\DeleteBulkAction::make(),
                ]),
            ]);
    }
```

CreatePost:

```
class CreatePost extends CreateRecord
{
    use \Filament\Resources\Pages\CreateRecord\Concerns\Translatable;

    protected static string $resource = PostResource::class;

    protected function getHeaderActions(): array
    {
        return [
            Actions\LocaleSwitcher::make(),
        ];
    }
}
```

EditPost:

```
class EditPost extends EditRecord
{
    use \Filament\Resources\Pages\EditRecord\Concerns\Translatable;

    protected static string $resource = PostResource::class;

    protected function getHeaderActions(): array
    {
        return [
            Actions\LocaleSwitcher::make(),
            Actions\DeleteAction::make(),
        ];
    }
}
```

ListPosts:

```
class ListPosts extends ListRecords
{
    use \Filament\Resources\Pages\ListRecords\Concerns\Translatable;

    protected static string $resource = PostResource::class;

    protected function getHeaderActions(): array
    {
        return [
            Actions\LocaleSwitcher::make(),
            Actions\CreateAction::make(),
        ];
    }
}
```

**Reply by gabyyro** [reply]
Did you manage to find a solution for this?

**Reply by christopheraseidl** [reply]
Sadly, no. It's on the backburner while I launch a couple of projects. I don't think I'll be looking at this any time soon, but I will return to it and provide an update when I do.

**Reply by gabyyro** [reply]
Thanks, likewise, have no solution yet.

**Reply by paul2Dev** [reply]
@christopheraseidl did you figured this one? i got the same problem :/

**Reply by christopheraseidl** [reply]
Hi Paul, I have not returned to this. However, Dan Harrin did address this a while back; [here's a copy of his response in another thread](https://github.com/filamentphp/filament/issues/11175#issuecomment-2103480974). He recommends a fork, https://github.com/lara-zeus/translatable. Best of luck!

**Reply by paul2Dev** [reply]
tnks @christopheraseidl i know the topic with the repo fork, i've used some code part from that repo but, i've tried the latest version from there and did not resolved my problem.

meanwhile i use a json column field on the same model instead of saving to releationship.
